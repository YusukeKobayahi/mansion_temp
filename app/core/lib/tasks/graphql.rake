# frozen_string_literal: true

module GraphqlTask
  extend Rake::DSL

  namespace :graphql do
    namespace :consumer do
      task schema_dump: :environment do
        require 'graphql/rake_task'

        GraphQL::RakeTask.new(
          load_schema: lambda { |_task|
            require File.expand_path('../../app/graphql/consumer/inno_schema', __dir__)
            Consumer::InnoSchema
          },
          directory: './graphql/consumer',
        )
        Rake::Task['graphql:schema:idl'].invoke
      end
    end
  end
end
