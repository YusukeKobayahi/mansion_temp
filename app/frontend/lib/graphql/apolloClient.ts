// 参考
// https://github.com/vercel/next.js/tree/canary/examples/with-apollo

import { useMemo } from "react";
import { ApolloClient, ApolloLink, InMemoryCache } from "@apollo/client";
import { createHttpLink } from "apollo-link-http";
import { from } from "apollo-link";
import { createFragmentArgumentLink } from "apollo-link-fragment-argument";
import { mergeDeep, relayStylePagination } from "@apollo/client/utilities";
import { isSSR } from "~/lib/utils";

/* eslint-disable @typescript-eslint/no-explicit-any -- キャッシュの型を事前に決められないため */
type CacheState = Record<string, any> | null;
/* eslint-enable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/naming-convention -- GraphQLのタイプ名 (このconventionに従わない)を指定したいので */
const createCache = () =>
  new InMemoryCache({
    typePolicies: {
      Mansion: {
        keyFields: ["uniqueCode"],
        fields: {
          salesHistories: relayStylePagination(),
        },
      },
    },
  });
/* eslint-enable @typescript-eslint/naming-convention */

const link = from([
  createFragmentArgumentLink(),
  createHttpLink({
    uri: `${process.env.NEXT_PUBLIC_CORE_API_URL}/consumer/graphql`,
    fetchOptions: {
      mode: "cors",
    },
  }),
]);

function createApolloClient() {
  return new ApolloClient({
    ssrMode: isSSR(),
    // See: https://github.com/apollographql/apollo-link/issues/1258#issuecomment-604031488
    // TODO: @apollo/clientバージョン上げたら直るかも
    link: (link as unknown) as ApolloLink,
    cache: createCache(),
    defaultOptions: {
      watchQuery: {
        // SSR時にuseQueryによってリクエストしてレンダリングする場合、
        // getServerSidePropsからキャッシュが渡っている必要がある。(でないと描画されない)
        // そのため、SSR時は常にcacheから取得されるべきで、実装意図を明確にするためPolicyも設定しておく
        // (エラーになるわけではない)
        fetchPolicy: isSSR() ? "cache-only" : "cache-first",
      },
    },
  });
}

let apolloClient: ReturnType<typeof createApolloClient>;

export function initializeApollo(
  initialState: CacheState = null
): ApolloClient<CacheState> {
  const _apolloClient = apolloClient ?? createApolloClient();

  // If your page has Next.js data fetching methods that use Apollo Client, the initial state
  // gets hydrated here
  if (initialState) {
    // Get existing cache, loaded during client side data fetching
    const existingCache = _apolloClient.extract();

    // Merge the existing cache into data passed from getStaticProps/getServerSideProps
    const data = mergeDeep(initialState, existingCache);

    // Restore the cache with the merged data
    _apolloClient.cache.restore(data);
  }
  // For SSG and SSR always create a new Apollo Client
  if (isSSR()) return _apolloClient;
  // Create the Apollo Client once in the client
  if (!apolloClient) apolloClient = _apolloClient;

  return _apolloClient;
}

export function useApollo(initialState: CacheState): ApolloClient<CacheState> {
  const store = useMemo(() => initializeApollo(initialState), [initialState]);
  return store;
}
