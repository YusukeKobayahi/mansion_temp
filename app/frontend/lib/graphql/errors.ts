export interface GraphQLError {
  graphQLErrors?: {
    message: string;
    extensions?: {
      code: string;
    };
  }[];
}

export const isNotFoundError = (err: GraphQLError): boolean => {
  const graphQLErrors = err.graphQLErrors;
  if (graphQLErrors == null) return false;

  return graphQLErrors[0]?.extensions?.code === "NOT_FOUND";
};
