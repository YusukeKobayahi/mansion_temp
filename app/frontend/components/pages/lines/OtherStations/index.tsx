import { LinesQuery } from "~/graphql/generated";

import styles from "~/components/pages/cities/OtherCities/index.module.scss";
import BoxHeading from "~/components/commons/headings/BoxHeading";
import LinkBlock from "~/components/commons/LinkBlock";

export interface OtherStationsProps {
  data: LinesQuery;
}

const OtherStations: React.FC<OtherStationsProps> = ({
  data,
}: OtherStationsProps) => {
  const items = data.lines.map((line) => {
    const links = line.stations.map((station) => (
      <LinkBlock
        key={station.id}
        name={station.name + "駅"}
        id={station.id}
        mansionCount={station.mansionCount}
        link="stations"
        eventParams={{
          action: "msl_area_click",
          category: "msl_ichiran",
          label: "msl_station",
          value: 1,
        }}
      />
    ));
    return (
      <div className={styles.items} key={line.id}>
        <h2>
          <BoxHeading>{line.name}の駅から中古マンションを探す</BoxHeading>
        </h2>
        <ul className={styles.list}>{links}</ul>
      </div>
    );
  });

  return <section className={styles.container}>{items}</section>;
};

export default OtherStations;
