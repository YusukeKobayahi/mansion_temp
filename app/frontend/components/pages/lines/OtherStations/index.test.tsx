import { render } from "@testing-library/react";
import mockData from "~/components/pages/lines/fixtures/mockData.json";
import { LinesQuery } from "~/graphql/generated";
import OtherStations from "~/components/pages/lines/OtherStations";

describe("components/pages/lines/OtherStations", () => {
  it("正常に表示される", () => {
    const mock = mockData as LinesQuery;
    const { asFragment } = render(<OtherStations data={mock} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
