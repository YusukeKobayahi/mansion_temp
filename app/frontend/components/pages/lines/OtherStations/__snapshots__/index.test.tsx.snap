// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/pages/lines/OtherStations 正常に表示される 1`] = `
<DocumentFragment>
  <section
    class="container"
  >
    <div
      class="items"
    >
      <h2>
        <div
          class="heading"
        >
          東海道線の駅から中古マンションを探す
        </div>
      </h2>
      <ul
        class="list"
      >
        <li
          class="item"
        >
          <button
            class="link"
            href="/stations/1320"
          >
            東京駅
            <span
              class="count"
            >
              (2件)
            </span>
          </button>
        </li>
        <li
          class="item"
        >
          <button
            class="link"
            href="/stations/1321"
          >
            新橋駅
            <span
              class="count"
            >
              (7件)
            </span>
          </button>
        </li>
        <li
          class="item"
        >
          <button
            class="link"
            href="/stations/1322"
          >
            品川駅
            <span
              class="count"
            >
              (29件)
            </span>
          </button>
        </li>
        <li
          class="item"
        >
          <button
            class="link"
            href="/stations/1323"
          >
            川崎駅
            <span
              class="count"
            >
              (82件)
            </span>
          </button>
        </li>
        <li
          class="item"
        >
          <button
            class="link"
            href="/stations/1324"
          >
            横浜駅
            <span
              class="count"
            >
              (43件)
            </span>
          </button>
        </li>
        <li
          class="item"
        >
          <button
            class="link"
            href="/stations/1325"
          >
            戸塚駅
            <span
              class="count"
            >
              (45件)
            </span>
          </button>
        </li>
        <li
          class="item"
        >
          <button
            class="link"
            href="/stations/1326"
          >
            大船駅
            <span
              class="count"
            >
              (21件)
            </span>
          </button>
        </li>
        <li
          class="item"
        >
          <button
            class="link"
            href="/stations/1327"
          >
            藤沢駅
            <span
              class="count"
            >
              (31件)
            </span>
          </button>
        </li>
        <li
          class="item"
        >
          <button
            class="link"
            href="/stations/1328"
          >
            辻堂駅
            <span
              class="count"
            >
              (32件)
            </span>
          </button>
        </li>
        <li
          class="item"
        >
          <button
            class="link"
            href="/stations/1329"
          >
            茅ヶ崎駅
            <span
              class="count"
            >
              (36件)
            </span>
          </button>
        </li>
        <li
          class="item"
        >
          <button
            class="link"
            href="/stations/1330"
          >
            平塚駅
            <span
              class="count"
            >
              (36件)
            </span>
          </button>
        </li>
        <li
          class="item"
        >
          <button
            class="link"
            href="/stations/1331"
          >
            大磯駅
            <span
              class="count"
            >
              (3件)
            </span>
          </button>
        </li>
        <li
          class="item"
        >
          <button
            class="link"
            href="/stations/1332"
          >
            二宮駅
            <span
              class="count"
            >
              (2件)
            </span>
          </button>
        </li>
        <li
          class="item"
        >
          <button
            class="link"
            href="/stations/1333"
          >
            国府津駅
            <span
              class="count"
            >
              (1件)
            </span>
          </button>
        </li>
        <li
          class="item"
        >
          <button
            class="link"
            href="/stations/1334"
          >
            鴨宮駅
            <span
              class="count"
            >
              (2件)
            </span>
          </button>
        </li>
        <li
          class="item"
        >
          <button
            class="link"
            href="/stations/1335"
          >
            小田原駅
            <span
              class="count"
            >
              (18件)
            </span>
          </button>
        </li>
      </ul>
    </div>
    <div
      class="items"
    >
      <h2>
        <div
          class="heading"
        >
          上野東京ラインの駅から中古マンションを探す
        </div>
      </h2>
      <ul
        class="list"
      >
        <li
          class="item"
        >
          <button
            class="link"
            href="/stations/2188"
          >
            東京駅
            <span
              class="count"
            >
              (2件)
            </span>
          </button>
        </li>
        <li
          class="item"
        >
          <button
            class="link"
            href="/stations/2189"
          >
            上野駅
            <span
              class="count"
            >
              (29件)
            </span>
          </button>
        </li>
      </ul>
    </div>
  </section>
</DocumentFragment>
`;
